### ISSUE TITLE
Track Email Conversation Threads with owasync.py

### ISSUE THREAD

**Issue by p6knewman** [linkedin/iris#274]
Hello,

When creating incidents via email messages, it would seem advantageous to only create an incident once per a conversation thread. Users will add the "On call" email address to a thread that requires emergency attention, which triggers the incident, but every successive message in that thread will continue to create new Iris incidents each time there is a reply.

I notice that there is a check in ResponseEmail.on_post for the 'In-Reply-To' header, but at least in my organization's Exchange implementation I don't see that header existing. However, Outlook/Exchange _does_ have a well-documented 'Thread-Index' header that tracks the uniqueness of a conversation thread:

https://msdn.microsoft.com/en-us/library/ee202481(v=exchg.80)?ranMID=24542&ranEAID=TnL5HPStwNw&ranSiteID=TnL5HPStwNw-go3ivEugoQ2BnSvLr55H8A&tduid=(753a0626561d1d154f69a0a528d10da2)(256380)(2459594)(TnL5HPStwNw-go3ivEugoQ2BnSvLr55H8A)()

Using the Thread-Index, I have some code that will store the unique Thread-Index ID in the incident context. When an incoming email message arrives that would normally create an incident, ResponseEmail.on_post will first check incidents created in the past X number of hours (configurable), and if the context already contains the same Thread-Index ID, skip creation of a new incident and log a warning.

Let me know what you think about the feature and I can send a PR if you like it. Thanks!

**Comment by jrgp** [comment]
Hey,

Thanks for looking into this. While we could store some metadata in the incident context and then iterate backward parsing the ID out of the json looking for it, this is likely to be inefficient and slow when you have **lots** of incidents.

We were hoping to find headers that only exist on email replies to a thread. Do you see any other headers aside from Thread-Index that we could merely check the presence of?

**Comment by jrgp** [comment]
Also, quick question: how are you pulling your emails from exchange? Using our OWA sync script or something else?

**Comment by p6knewman** [comment]
I agree that looking backwards in *all* incident context carries the risk of being very slow with a lot of incidents. For this reason, I'd restrict it to incidents within the last 24 hours by default (with a configurable lookback time). As long as the incident.created is indexed (and it is), I don't foresee this being too much of a drag on performance.

Can I show you what I have in mind? Here's a diff of the proposed changes:
https://github.com/linkedin/iris/compare/master...p6knewman:issue_274_track_thread_index

**Comment by jrgp** [comment]
Hey,

This might cause problems for us and other very high volume iris instances as we have around 2000 incidents per day.

It would be cleanest for this process to not have state. Would it be possible for you to instead of CC'ing, BCC iris when you need to loop it into a thread and that way subsequent replies won't create incidents?

Regards

**Comment by p6knewman** [comment]
Not sure I understand, or maybe we're not on the same page. The query of recent incidents is this line:
https://github.com/linkedin/iris/compare/master...p6knewman:issue_274_track_thread_index#diff-ddac0526a1fb66b4b38d3eaa27bce478R3157

If the worst-case scenario for incident volume is on the order of 2000 per/day, the WHERE clause on the default configuration setting would only produce 2000 rows returned. That doesn't seem too onerous to work with, IMO. If you have a high volume DB, would you be willing to pull down my branch and test performance with a couple test incidents created via email? I could also work on creating a ton of junk data on my dev instance to see what performance is like.

As for BCC vs. CC, that unfortunately wouldn't solve my use case. The users that may/will be sending incidents in via email may are unlikely to remember to go through such specific steps when logging an incident.

**Comment by p6knewman** [comment]
I went ahead and generated some junk data on my dev instance. I generated 2000 incidents per day going back for the last 365 days (730k total). I timed the performance of both the MySQL query and the combined time it takes to query the DB + ujson.loads the context and iterate over it looking for the same thread_id.

The performance numbers were quite acceptable even up to a retrieval of ~28,000 incidents (~336hrs at 2000 incidents/day), which really seems like a silly long period to be looking back for duplicate incidents. Long before that far back, I'd be considering this a new incident and it's time to page someone out again. At a more reasonable period of 24 hours, the performance impact is roughly 90-100ms. 

```
with 730k incidents of random context data, 2000 incidents per a day, incident_lookback_time 48 hours:
web.1    | [2017-10-05 09:05:38 +0000] [98023] [WARNING] iris.api Fetched 4003 incidents from DB in 0.09 seconds
web.1    | [2017-10-05 09:05:38 +0000] [98023] [WARNING] iris.api Total time to fetch from DB + parse 4003 recent incidents: 0.1 seconds
web.1    | 127.0.0.1 - - [05/Oct/2017:09:05:38 -0500] "POST /v0/response/email HTTP/1.1" 204 - "-" "python-requests/2.9.1"


incident_lookback_time 72 hours:
web.1    | [2017-10-05 09:06:38 +0000] [98742] [WARNING] iris.api Fetched 6003 incidents from DB in 0.12 seconds
web.1    | [2017-10-05 09:06:38 +0000] [98742] [WARNING] iris.api Total time to fetch from DB + parse 6003 recent incidents: 0.13 seconds
web.1    | 127.0.0.1 - - [05/Oct/2017:09:06:38 -0500] "POST /v0/response/email HTTP/1.1" 204 - "-" "python-requests/2.9.1"


incident_lookback_time 144 hours:
web.1    | [2017-10-05 09:07:45 +0000] [99102] [WARNING] iris.api Fetched 12001 incidents from DB in 0.3 seconds
web.1    | [2017-10-05 09:07:45 +0000] [99102] [WARNING] iris.api Total time to fetch from DB + parse 12001 recent incidents: 0.35 seconds
web.1    | 127.0.0.1 - - [05/Oct/2017:09:07:45 -0500] "POST /v0/response/email HTTP/1.1" 204 - "-" "python-requests/2.9.1"


incident_lookback_time 336 hours:
web.1    | [2017-10-05 09:09:28 +0000] [99599] [WARNING] iris.api Fetched 27994 incidents from DB in 0.49 seconds
web.1    | [2017-10-05 09:09:28 +0000] [99599] [WARNING] iris.api Total time to fetch from DB + parse 27994 recent incidents: 0.54 seconds
web.1    | 127.0.0.1 - - [05/Oct/2017:09:09:28 -0500] "POST /v0/response/email HTTP/1.1" 204 - "-" "python-requests/2.9.1"
```

**Comment by p6knewman** [comment]
@jrgp, any thoughts on this? If preferred, I could make this whole feature subject to an on/off config option and default it to off.
